clang/lib/Sema/SemaTemplateInstantiateDecl.cpp

```diff
@@ -6168,11 +6168,12 @@ static void processSYCLKernel(Sema &S, FunctionDecl *FD, MangleContext &MC) {
   if (S.LangOpts.SYCLIsDevice) {
     S.ConstructOpenCLKernel(FD, MC);
   } else if (S.LangOpts.SYCLIsHost) {
-    CXXRecordDecl *CRD = (*FD->param_begin())->getType()->getAsCXXRecordDecl();
+    S.ConstructOpenCLKernel(FD, MC);
+    /*CXXRecordDecl *CRD = (*FD->param_begin())->getType()->getAsCXXRecordDecl();
```

Comment:
My current implementation causes assertion here. I have commented out. Any suggestion would be greatly appreciated. We need to handle same test cases on host as device.

Reply:
I think this is because the Kernel is now passed as a reference.  Try this:

```
-    CXXRecordDecl *CRD = (*FD->param_begin())->getType()->getAsCXXRecordDecl();
+    QualType KernelParamTy = (*FD->param_begin())->getType();
+    const CXXRecordDecl *CRD;
+    if (KernelParamTy->isReferenceType())
+      CRD = KernelParamTy->getPointeeCXXRecordDecl();
+    else
+      CRD = KernelParamTy->getAsCXXRecordDecl();

```

Reply:
Thanks @premanandrao for the suggestion. I will take a look at this.